Add multi-class classification target

giovannt0 opened the pull request.

kirschte committed:
extended for multi-class classification

kirschte committed:
added example task for classification

giovannt0 commented on estimator.py:
```diff
@@ -21,6 +21,7 @@ def __init__(self,
                nb_trees_per_ensemble: int,
                max_depth: int,
                learning_rate: float,
+               n_classes: int = None,
```
Can you type the argument as `Optional[int]` for consistency?

giovannt0 commented on estimator.py:
```diff
@@ -109,16 +112,20 @@ def predict(self, X: np.array) -> np.array:
-        class_preds.append(1)
-    return class_preds
+    try:
+      preds = self.model.PredictLabels(X)
```
You can probably directly return `self.model.PredictLabels(X)` or `self.model.Predict(X)` rather than assigning the result to the variable `preds`, since it's unused afterwards

giovannt0 commented on estimator.py:
```diff
@@ -109,16 +112,20 @@ def predict(self, X: np.array) -> np.array:
+    except ValueError:
+      preds = self.model.Predict(X)
+    return preds
+#    if not self.binary_classification:
```
Delete commented lines

giovannt0 commented on model.py:
```diff
@@ -39,6 +40,7 @@ class GradientBoostingEnsemble:
   def __init__(self,
                nb_trees: int,
                nb_trees_per_ensemble: int,
+               n_classes: int = None,
```
`Optional[int] = None`

giovannt0 commented on model.py:
```diff
@@ -125,29 +132,27 @@ def Train(self,
+                      (1 - self.learning_rate), tree_index))
+
+      # Privacy budget allocated to each tree
+      tree_privacy_budget = np.divide(self.privacy_budget, nb_ensembles * self.loss_.K)
```
Why is `nb_ensembles` multiplied by `self.loss_.K` here?

giovannt0 commented on model.py:
```diff
@@ -125,29 +132,27 @@ def Train(self,
-        if tree_index > 0:
-          gradients = self.ComputeGradientForLossFunction(
-              y_ensemble, self.Predict(X_ensemble))
+        # gradient initialization will happen later in the per-class-loop
```
I see here you removed the gradients initialisation. Does it happen in the per-class-loop when using regression as well? If not, please add it back and adapt the `if` statement to work accordingly.

kirschte replied:
Regression output remains the same as before. :+1:

giovannt0 replied:
This `if` statement only is reached when there are multiple ensembles (not the case in `example.py`). Can you check if with e.g. 100 trees and 20 trees per ensemble there's no impact?
Also, irrespective of the output, why shouldn't we reinitialise the gradients here? Since these trees are independent from the previous ones and work on different data I'd expect the need of initialising the gradients again (note that the choice of initialisation is another topic, I'm not sure which is best in this case as this is not mentioned in the original paper)

kirschte replied:
Regression output remains the same even with multiple ensembles. :)
In fact, the gradients are reinitialized, just later on in [L201-203](https://github.com/giovannt0/dpgbdt/pull/1/files/1f66c420a24a00e3cbe7840aa3a680778c3047e6#diff-fada037ad086638e65c7ae77e3d223963e9afaa26326aab0ea718f4013176e43R201-R203). Sorry for the confusion, I just had to reorganize some code as the gradient initialization has to be done within the per-class loop and some constant parts shouldn't be.
Anyhow, it is a nice functionality to run multiple ensembles and imho it should still work ;)

giovannt0 commented on model.py:
```diff
@@ -185,57 +190,65 @@ def Train(self,
-      rmse = np.sqrt(np.mean(np.square(y_test - self.Predict(X_test))))
-      if rmse >= prev_rmse:
+      k_trees = []
+      for kth_tree in range(self.loss_.K):
```
What is this loop for? Is it iterating over all trees? Something else? How different is it compared to the loop [here](https://github.com/giovannt0/dpgbdt/pull/1/files#diff-fada037ad086638e65c7ae77e3d223963e9afaa26326aab0ea718f4013176e43R138)? What is the expected behaviour for the regression task vs. the classification one? Perhaps add a few comments to explain what's happening.

giovannt0 commented on model.py:
```diff
@@ -185,57 +190,65 @@ def Train(self,
+            cat_idx=self.cat_idx,
+            num_idx=self.num_idx)
+        tree.Fit(X_tree,
+                 (y_ensemble == kth_tree).astype(np.float64) if self.loss_.is_multi_class else y_ensemble,
```
Maybe `(y_ensemble == kth_tree).astype(np.float64) if self.loss_.is_multi_class else y_ensemble` can be its own statement rather than within the function call for more readability?

giovannt0 commented on model.py:
```diff
@@ -246,74 +259,77 @@ def Train(self,
-            if not candidate.root_node.X.any():
+    for index, k_three_tree in enumerate(k_trees):
+      k_trees_ = []
+      for k, three_tree in enumerate(k_three_tree):
```
Why is there an additional loop here now? Is that only for the classification task? Maybe add a comment explaining what's happening

giovannt0 commented on model.py:
```diff
@@ -345,10 +362,23 @@ def SplitNode(self,
-        node.gradients[lhs]) + self.l2_lambda))  # type: float
-    rhs_prediction = (-1 * np.sum(node.gradients[rhs]) / (len(
-        node.gradients[rhs]) + self.l2_lambda))  # type: float
+    if self.loss_.is_multi_class and len(node.gradients[lhs]) > 0:
```
Here I think we should introduce a function to make it a little easier to read. Something like `ComputePredictions(node: DecisionNode, indices: List[int], is_multi_class: bool)`. Then we can just call:

```
rhs_predictions = self.ComputePredictions(node, rhs, self.loss_.is_multi_class)
lhs_predictions = self.ComputePredictions(node, lhs, self.loss_.is_multi_class)
```

Makes it a little more readable

kirschte replied:
Great idea! There is actually another occurrence of this snipped; all of which I've now integrated into one function.

giovannt0 commented on model.py:
```diff
@@ -392,23 +424,41 @@ def Predict(self, X: np.array) -> np.array:
       np.array: The predictions.
     """
-    predictions = np.sum(tree.Predict(X) for tree in self.trees)
+    predictions = np.sum([[tree.Predict(X) for tree in k_trees] for k_trees in self.trees], axis=0).T
```
Is `k_trees` an ensemble of tree per label to predict here? In the regression task shouldn't `predictions` be a `np.ndarray` rather than a list of `np.ndarray`?

kirschte replied:
Per se your're right. However as `Predict` is also used internally, it is okay to abstract the regression onto a one-class regression (thus a list containing only one array).
In the user-friendly interface `DPGBDT` in `estimator.py` though, I now converted it back to its original shape, an `np.ndarray`.

giovannt0 replied:
Sounds good, thanks for the clarification!

giovannt0 commented on model.py:
```diff
@@ -392,23 +424,41 @@ def Predict(self, X: np.array) -> np.array:
-    return np.multiply(-1, np.subtract(y, y_pred))
+    if self.loss_.is_multi_class:
+      y = (y == k).astype(np.float64)
+    return -self.loss_.negative_gradient(y, y_pred, k=k)
```
I'm not familiar with Scikit's `negative_gradient` method here; but if it's negative already as the name implies, do we need the extra `-` sign in front of `self.loss_`?

giovannt0 commented on model.py:
```diff
@@ -830,17 +900,23 @@ def FindBestSplit(self,
-    prediction = (-1 * np.sum(gradients) / (len(
-        gradients) + self.l2_lambda))  # type: float
+    if self.loss.is_multi_class and len(gradients) > 0:
+      prediction = -1 * np.sum(gradients) * (self.loss.K - 1) / self.loss.K
```
Can you link here a paper / documentation that explains the reasoning for the predictions formula? Just for reference.

giovannt0 reviewed (commented):
A few comments / questions. Otherwise it's looking good!

kirschte committed:
addresses comments + improved classification quality

giovannt0 commented on estimator.py:
```diff
@@ -112,20 +112,18 @@ def predict(self, X: np.array) -> np.array:
     """
     assert self.model
+    # try (multi-class) classification output first,
+    # ow fallback to the raw regression values
```
Small typo here: `ow` -> `or`

kirschte replied:
actually, ow means otherwise^^

giovannt0 commented on estimator.py:
```diff
@@ -112,20 +112,18 @@ def predict(self, X: np.array) -> np.array:
+      if not self.binary_classification:
+        return reg_preds
+      else:
+        return np.where(reg_preds < 0, -1, 1)
```
Sweet, thanks!

giovannt0 commented on model.py:
```diff
@@ -231,8 +234,11 @@ def Train(self,
+        # as each tree is a per-class regressor
         tree.Fit(X_tree,
-                 (y_ensemble == kth_tree).astype(np.float64) if self.loss_.is_multi_class else y_ensemble,
+                 (y_ensemble == kth_tree).astype(np.float64) if self.loss_.is_multi_class
```
Maybe (y_ensemble == kth_tree).astype(np.float64) if self.loss_.is_multi_class else y_ensemble can be its own statement a few lines before rather than within the function call for more readability?

giovannt0 reviewed (commented):
A couple questions left, otherwise LGTM; thanks!

kirschte committed:
some minor beautification

kirschte committed:
privacy considerations around how to handle multiple trees per ensemble

Commit:
Update model.py

updated multi-class loss function according to the original paper, which now gives the expected results while using the proper neg. gradients.
sklearn uses a different version (i.e. eq 33) instead, but – for whatever reason – the latter doesn't work properly on this codebase…

Commit:
minor doc fix

kirschte committed:
sometimes gradients weren't properly updated… Thus multi-class leaf prediction could be rolled back to previous formulation
